### PR TITLE
fix: bump Socket.IO maxHttpBufferSize + decode non-image file attachments

### PR DESCRIPTION
**Pull request by Pizzaface**
### 1. Socket.IO maxHttpBufferSize (100 MB)

PizzaPi didn't set `maxHttpBufferSize` on its Socket.IO server, inheriting the 1 MB default. Large session state (e.g. conversations with embedded screenshots, observed at ~17 MB) caused instant `transport close` on every connection — sessions became unreachable with no error.

**Fix:** Bump `maxHttpBufferSize` to 100 MB as a safety valve.

### 2. Non-image file attachments decoded inline

When users attached text-based files (JSON, Python, Markdown, etc.) from the web UI, the content was silently dropped — replaced with a useless placeholder like `[Attachment provided by web client: file.json]`. The base64 data was fetched but discarded for anything that wasn't `image/*`.

**Fix:** Added `isTextMimeType()` helper that detects text files by MIME type (`text/*`, `application/json`, `application/yaml`, etc.) with filename extension fallback. Three-way branching in `buildUserMessageFromRemoteInput()`:
- **Image** → image content part (unchanged)
- **Text file** → base64-decoded, included inline wrapped in `--- filename --- / --- end filename ---`
- **Binary non-image** → honest placeholder noting content was not included

### Tests

- Expanded `remote-input.test.ts` from 7 → 20 tests covering `isTextMimeType()` and all attachment type combinations
- All 216 tests pass, typecheck clean

## Future improvements (tracked in Godmother)

- Strip or compress base64 image data from session state before broadcast
- Implement state pagination / lazy loading for conversation history

### CODE REVIEW

File: packages/server/src/index.ts [hunk 1]
```diff
@@ -132,6 +132,12 @@ try {
             origin: getTrustedOrigins(),
             credentials: true,
         },
+        // Socket.IO default maxHttpBufferSize is 1 MB, which silently drops
+        // connections when a session state payload exceeds it (e.g., sessions
+        // with embedded screenshots can easily reach 10–50 MB). The transport
+        // close happens with no error surfaced to the user. Bump to 100 MB
+        // as a safety valve so large sessions remain accessible.
+        maxHttpBufferSize: 100 * 1024 * 1024, // 100 MB
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/index.ts:149]: Valid concern, but this is pre-existing behavior — the default `/` namespace has always been unguarded regardless of `maxHttpBufferSize`. All meaningful namespaces (`/relay`, `/runner`, `/viewer`, `/terminal`, `/hub`) have auth middleware.

The 100 MB limit is needed because runners send raw session state (with inline screenshots) before server-side image stripping. This PR actually *reduces* the sustained memory risk by extracting images server-side.

Filed a follow-up issue for adding a global connection guard / rate limiter as defense-in-depth.

File: packages/server/src/ws/strip-images.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        const mimeType = typeof b.mimeType === "string"
+            ? b.mimeType
+            : typeof source?.media_type === "string"
+                ? source.media_type
+                : typeof source?.mediaType === "string"
+                    ? source.mediaType
+                    : "image/png";
+
+        const attachmentId = crypto.randomUUID();
+        extracted.push({ attachmentId, mimeType, base64Data: data, sizeBytes });
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/ws/strip-images.ts]: Good catch — fixed. `extractImages()` now uses a SHA-256 content hash (truncated to 24 hex chars) as the attachment ID instead of `crypto.randomUUID()`. Same image content → same ID, so repeated state updates are deduplicated.

`storeExtractedImage()` also now short-circuits when the attachment ID already exists in the store — it just refreshes the expiry TTL and returns the existing record.

Added tests for deterministic IDs.

File: packages/server/src/attachments/store.ts [hunk 3]
```diff
@@ -95,6 +97,62 @@ export async function deleteStoredAttachment(attachmentId: string): Promise<void
...
+    attachmentId: string;
+    sessionId: string;
+    ownerUserId: string;
+    mimeType: string;
+    base64Data: string;
+}): Promise<StoredAttachment> {
+    const { attachmentId, sessionId, ownerUserId, mimeType, base64Data } = input;
+
+    // Decode base64 to raw bytes
+    const bytes = Buffer.from(base64Data, "base64");
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/attachments/store.ts]: Good catch — fixed. `storeExtractedImage()` now strips the `data:...;base64,` prefix before calling `Buffer.from(..., "base64")`.

Also exported a `stripDataUriPrefix()` helper from `strip-images.ts` for reuse, with tests covering various MIME types and edge cases.

File: packages/server/src/attachments/store.ts [hunk 4]
```diff
@@ -95,6 +97,76 @@ export async function deleteStoredAttachment(attachmentId: string): Promise<void
...
+    base64Data: string;
+}): Promise<StoredAttachment> {
+    const { attachmentId, sessionId, ownerUserId, mimeType, base64Data } = input;
+
+    // Deduplicate: if this content-hashed ID already exists, just refresh its
+    // expiry and return the existing record — no need to re-decode/write.
+    const existing = attachments.get(attachmentId);
+    if (existing) {
+        const refreshedExpiry = Date.now() + EXTRACTED_IMAGE_TTL_MS;
+        existing.expiresAt = new Date(refreshedExpiry).toISOString();
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/attachments/store.ts:152]: Already handled — `contentHash()` in `strip-images.ts` includes `userId` in the hash:

```ts
function contentHash(data: string, userId: string): string {
    return createHash("sha256").update(userId).update(":").update(data).digest("hex").slice(0, 24);
}
```

Different users producing identical images get different `attachmentId` values, so the dedup in `storeExtractedImage()` never crosses user boundaries. The `ownerUserId` enforcement on download is a defense-in-depth layer that remains intact.

File: packages/server/src/ws/sio-registry.ts [hunk 5]
```diff
@@ -490,7 +504,7 @@ export async function updateSessionState(sessionId: string, state: unknown): Pro
         );
     }
 
-    void recordRelaySessionState(sessionId, state).catch((error) => {
+    void recordRelaySessionState(sessionId, strippedState).catch((error) => {
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/ws/sio-registry.ts:507]: Fixed — extracted attachment metadata is now persisted to SQLite (`extracted_attachment` table) and rehydrated on server startup.

- `storeExtractedImage()` writes to SQLite on store and on TTL refresh
- `deleteStoredAttachment()` removes from SQLite
- `rehydrateExtractedAttachments()` runs at startup: cleans expired rows, verifies files exist on disk, repopulates in-memory Map
- Table created via `ensureExtractedAttachmentTable()` in migrations

After a restart, persisted session state URLs now resolve correctly because the attachment registry is rehydrated from the database.

[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/ws/sio-registry.ts:507]: Same fix as the sibling thread — extracted attachment metadata is now persisted to SQLite and rehydrated on startup. Image URLs in durable session snapshots survive server restarts.

File: packages/server/src/attachments/store.ts [hunk 6]
```diff
@@ -93,6 +96,83 @@ export async function deleteStoredAttachment(attachmentId: string): Promise<void
...
+    const bytes = Buffer.from(rawB64, "base64");
+    const ext = mimeType.split("/").pop() ?? "png";
+    const filename = `extracted-${attachmentId}.${ext}`;
+    const targetPath = path.join(uploadRoot, filename);
+
+    await mkdir(uploadRoot, { recursive: true });
+    await Bun.write(targetPath, bytes);
+
+    const createdAtMs = Date.now();
+    const expiresAtMs = createdAtMs + EXTRACTED_IMAGE_TTL_MS;
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/attachments/store.ts:180]: Fixed — `sweepExpiredAttachments()` and `rehydrateExtractedAttachments()` now check `getDurableSessionIds()` before deleting expired extracted attachments. If the attachment belongs to a pinned or non-ephemeral session that is still alive, the TTL is renewed (another 24h) instead of deleting. This means extracted images for long-lived sessions are kept indefinitely as long as the session itself persists.

File: packages/server/src/attachments/store.ts [hunk 7]
```diff
@@ -104,3 +184,106 @@ export async function sweepExpiredAttachments(nowMs: number = Date.now()): Promi
...
+ */
+export async function rehydrateExtractedAttachments(): Promise<number> {
+    const nowMs = Date.now();
+    const nowIso = new Date(nowMs).toISOString();
+
+    // Delete expired rows first
+    await getKysely()
+        .deleteFrom("extracted_attachment")
+        .where("expiresAt", "<=", nowIso)
+        .execute();
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/attachments/store.ts]: Fixed — startup rehydration now fetches expired rows first, deletes their on-disk files via `rm(filePath)`, then removes from SQLite. Additionally, rows belonging to durable sessions are renewed rather than deleted, so no orphans are created for long-lived sessions either.

File: packages/server/src/attachments/store.ts [hunk 8]
```diff
@@ -93,14 +96,257 @@ export async function deleteStoredAttachment(attachmentId: string): Promise<void
...
+}): Promise<StoredAttachment> {
+    const { attachmentId, sessionId, ownerUserId, mimeType, base64Data } = input;
+
+    // Deduplicate: if this content-hashed ID already exists, just refresh its
+    // expiry and return the existing record — no need to re-decode/write.
+    const existing = attachments.get(attachmentId);
+    if (existing) {
+        const refreshedExpiry = Date.now() + EXTRACTED_IMAGE_TTL_MS;
+        existing.expiresAt = new Date(refreshedExpiry).toISOString();
+        existing.expiresAtMs = refreshedExpiry;
```
[the remark this answers is not in the document]
Comment (Pizzaface) [reply packages/server/src/attachments/store.ts:153]: Fixed — the dedup path now updates `existing.sessionId = sessionId` before refreshing the TTL and persisting. This ensures the attachment tracks the most recent session referencing it, so sweep/rehydration durability checks use the latest session's lifecycle rather than the first one's.

**Comment by Pizzaface** [comment]
**Re: P1 — Track all sessions referencing deduplicated extracted images (store.ts:130)**

Fixed — extracted attachments now track ALL session IDs that reference them, not just the latest one:

- Added an `extracted_attachment_session` junction table in SQLite (with unique index)
- In-memory `Map<attachmentId, Set<sessionId>>` parallels the DB
- `storeExtractedImage()` adds to the set on both new records and dedup hits
- `sweepExpiredAttachments()` and `rehydrateExtractedAttachments()` now check if ANY referencing session is durable before deleting
- Deletion cleans up the junction table alongside the main record

Also fixed `contentHash()` in `strip-images.ts` to normalize data URI prefixes before hashing, so the same image bytes produce the same attachment ID regardless of encoding form (raw base64 vs `data:...;base64,...`).

---

**Re: P2 — Await attachment saves to prevent same-name overwrite races (remote-input.ts:148)**

Fixed — `saveSessionAttachment()` is now awaited instead of fire-and-forget. This serializes saves within a single message's attachment loop, preventing the non-atomic read-then-write race in `deduplicateFilename()`.

**Comment by Pizzaface** [comment]
**Re: P1 — Make filename dedup atomic across concurrent saves (session-attachments.ts:109)**

Fixed — added a per-session promise-chain mutex (`withSessionLock`) that serializes the `readdir → dedup → write` sequence. Concurrent saves to the same session now queue up; saves to different sessions remain parallel. Tests confirm 3 concurrent saves of `file.txt` produce 3 unique files (`file.txt`, `file-2.txt`, `file-3.txt`).

---

**Re: P2 — Clean up persisted session attachments on session teardown (remote-input.ts:152)**

Rather than just hooking `cleanupSessionAttachments()` to session end (which misses crashes/restarts), we went with a **two-layer approach**:

1. **Prompt cleanup**: `cleanupSessionAttachments()` is now called in both the `session_ended` and `kill_session` handlers in the runner daemon — handles normal teardown.

2. **Orphan sweep**: Added `sweepOrphanedAttachments(activeSessionIds)` that cross-references attachment directories against the daemon's `runningSessions` map. This runs on `runner_registered` (after re-adopting sessions from a restart), catching any dirs orphaned by crashes or unclean shutdowns.

Also fixed a minor leak where `sessionLocks` entries weren't cleaned up on session teardown.

**Comment by Pizzaface** [comment]
**Re: P2 — Clean attachments even when session record is already gone (daemon.ts:718)**

Fixed — the root cause was that `child.on("exit")` fires on normal exit and removes the entry from `runningSessions` *before* the relay sends `session_ended`, so the `if (entry)` guard in `session_ended` was never satisfied in the common case.

Two-part fix:
1. Added `cleanupSessionAttachments` directly to `child.on("exit")` — this covers all newly-spawned sessions on normal and abnormal exit.
2. Moved `cleanupSessionAttachments` outside the `if (entry)` guard in `session_ended` — this covers adopted sessions (no `child.on("exit")` handler) and acts as a belt-and-suspenders backup for newly-spawned sessions.

---

**Re: P2 — Keep fallback text when attachment bytes are missing (remote-input.ts:175)**

Fixed — added an `else` branch so when a text attachment is identified but both `savedPath` and `dataBase64` are absent (relay fetch failed, invalid data URL, etc.), a visible placeholder `[Attached file: {label} — content unavailable]` is pushed into `parts` instead of silently dropping the attachment. This ensures the outgoing user message is never silently emptied due to a failed text attachment load.

**Comment by Pizzaface** [comment]
**Re: P2 — Persist extracted attachment metadata before returning (store.ts:185)**

Fixed — `storeExtractedImage()` now `await`s both `persistExtractedAttachment()` and `persistSessionRef()` before returning. The file write and the SQLite record are now durably committed before the caller can store the `/api/attachments/:id` URL in session state, closing the crash window that caused 404s on rehydration.

---

**Re: P2 — Compare orphan sweep against sanitized session IDs (session-attachments.ts:215)**

Fixed — `sweepOrphanedAttachments()` now builds a sanitized-ID set (`[...activeSessionIds].map(id => sanitize(id))`) before comparing against directory names. Previously, session IDs containing `/` or `:` would never match their already-sanitized directory names and the live directory would be falsely deleted. Added a test: raw ID `"runner/abc:1"` (stored as `"runner_abc_1"`) is now correctly preserved.

**Comment by Pizzaface** [comment]
**Re: restart guard race — session_ended beating child.on("exit")**

Fixed. The root cause: `restartingSessions.add(sessionId)` was inside `child.on("exit")`, but the relay emits `session_ended` the moment the old relay socket disconnects — over Socket.IO — and that network event can arrive at the daemon before the local OS exit notification fires. When it did, `session_ended` saw an empty `restartingSessions` and took the teardown path, deleting attachments for a session that was mid-restart.

**Fix**: workers are now spawned with an IPC channel (`stdio: ["ignore", "inherit", "inherit", "ipc"]` — same pattern already used by the terminal worker). Before calling `process.exit(43)`, the worker sends `process.send?.({ type: "pre_restart" })`. The daemon's `child.on("message")` handler adds the session to `restartingSessions` synchronously — before the process exits and before any relay event can arrive. The existing `child.on("exit")` add is kept as a belt-and-suspenders fallback.

The optional chaining on `process.send?.(...)` makes this a no-op in non-daemon contexts (direct invocation without a parent IPC channel).

Separately filed Godmother idea Q88RyiEr for a proper daemon test harness to cover restart lifecycle and other session scenarios end-to-end.

**Comment by Pizzaface** [comment]
**Re: P2 — Set restart guard before relay can emit session_ended (daemon.ts:1677)**

Already addressed by the IPC mechanism at line 1661. The worker sends `{ type: "pre_restart" }` via `process.send()` over the fd[3] IPC pipe *before* calling `process.exit(43)` (see `remote-exec-handler.ts:417`). The daemon's `child.on("message")` handler adds to `restartingSessions` synchronously on receipt.

Since the IPC message travels over a local Unix domain socket pipe while `session_ended` requires a TCP round-trip through Socket.IO (worker → server → runner), the IPC message always arrives first. The `restartingSessions.add()` on line 1677 inside `child.on("exit")` is documented as a belt-and-suspenders fallback.

---

**Re: P1 — Preserve durable extracted attachments on read path (store.ts:212)**

Good catch — fixed in beda004. `getStoredAttachment()` now checks whether an expired extracted image belongs to a durable (pinned/non-ephemeral) session before deleting. The read path mirrors the sweep logic: renew TTL for durable sessions, delete for ephemeral ones. Made the function async to support the `getDurableSessionIds()` query.

**Comment by Pizzaface** [comment]
**Re: P1 — Skip attachment teardown on relay reconnections (daemon.ts:739)**

Fixed in f10dde7. The server now includes `reason` in the `session_ended` event payload (`sio-registry.ts`). The daemon checks for `reason === "Session reconnected"` and skips both the `runningSessions` deletion and attachment cleanup — the worker is still alive during relay reconnections and will re-register immediately.

**Comment by Pizzaface** [comment]
**Re: P2 — Await dedupe-path attachment metadata writes (store.ts:162)**

Fixed in 0b0535d. The dedupe branch now `await`s both `persistExtractedAttachment()` and `persistSessionRef()` before returning, matching the non-dedupe path. A crash in the fire-and-forget window can no longer leave stale session linkage.
